lib/server.go

```diff
@@ -1738,16 +1842,36 @@ func (srv *Server) _addNewTxn(
 	}
 
 	srv.blockchain.ChainLock.RLock()
-	newlyAcceptedTxns, err := srv.mempool.ProcessTransaction(
-		txn, true /*allowUnconnectedTxn*/, rateLimit, peerID, verifySignatures)
+	tipHeight := uint64(srv.blockchain.BlockTip().Height)
 	srv.blockchain.ChainLock.RUnlock()
-	if err != nil {
-		return nil, errors.Wrapf(err, "Server._handleTransaction: Problem adding transaction to mempool: ")
+
+	// Only attempt to add the transaction to the PoW mempool if we're on the
+	// PoW protocol. If we're on the PoW protocol, then we use the PoW mempool's,
+	// txn validity checks to signal whether the txn has been added or not. The PoW
+	// mempool has stricter txn validity checks than the PoW mempool, so this works
+	// out conveniently, as it allows us to always add a txn to the PoS mempool.
+	if srv.params.IsPoWBlockHeight(tipHeight) {
+		_, err := srv.mempool.ProcessTransaction(
+			txn, true /*allowUnconnectedTxn*/, rateLimit, peerID, verifySignatures)
+		if err != nil {
+			return nil, errors.Wrapf(err, "Server._addNewTxn: Problem adding transaction to mempool: ")
+		}
+
+		glog.V(1).Infof("Server._addNewTxn: newly accepted txn: %v, Peer: %v", txn, pp)
 	}
 
-	glog.V(1).Infof("Server._addNewTxnAndRelay: newlyAcceptedTxns: %v, Peer: %v", newlyAcceptedTxns, pp)
+	// Always add the txn to the PoS mempool. This should always succeed if the txn
+	// addition into the PoW mempool succeeded above.
+	mempoolTxn := NewMempoolTransaction(txn, uint64(time.Now().UnixMicro()))
+	if err := srv.posMempool.AddTransaction(mempoolTxn, true /*verifySignatures*/); err != nil {
+		return nil, errors.Wrapf(err, "Server._addNewTxn: problem adding txn to pos mempool")
+	}
 
-	return newlyAcceptedTxns, nil
+	// At this point, we know the transaction has been run through the mempool.
+	// Now wait for an update of the ReadOnlyUtxoView so we don't break anything.
+	srv.GetMempool().BlockUntilReadOnlyViewRegenerated()
```

Comment:
I don't think you want BlockUntilReadOnlyViewRegenerated here. I think _addTxn is called when we initialize the mempool, which means you'll be blasting it with txns. Instead, you want to do this wait in BroadcastTransaction() only, which is where it was before anyway. You can just look at the diff and make sure that the position of BlockUntilReadOnlyViewRegenerated hasn't changed, and make sure it makes sense why it hasn't changed.

Reply:
Done